time-machine/main.py

```diff
@@ -0,0 +1,138 @@
+        username=os.getenv('VISIER_USERNAME'),
+        password=os.getenv('VISIER_PASSWORD'),
+        client_id=os.getenv('VISIER_CLIENT_ID'),
+        client_secret=os.getenv('VISIER_CLIENT_SECRET'),
+        redirect_uri=os.getenv('VISIER_REDIRECT_URI'),
+        vanity=os.getenv('VISIER_VANITY'))
+    return config
+
+
+def parse_args():
+    logger.info(f"Parsing command line arguments: {sys.argv[1:]}.")  # Add arguments to log
+    parser = argparse.ArgumentParser(description="Time Machine sample.")
+    parser.add_argument('-q', '--query_file_path', type=str, required=True,
+                        help="File path to aggregate query definition.")
+    parsed_args = parser.parse_args()
+    return parsed_args
+
+
+def get_temp_file_path(args):
+    file_name, _ = os.path.splitext(os.path.basename(args.query_file_path))
+    storage_path = os.getenv('STORAGE_PATH', default='tmp_storage')
+    date_timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
+    temp_data_file = os.path.join(storage_path, f'{file_name}_{date_timestamp}.csv')
+    return temp_data_file
+
+
+def download_data(config: Configuration,
+                  query_file_path: str,
+                  data_file_path: str):
+    """
+    Downloads metrics data and saves it to a file.
+    Args:
+        config: The API configuration object.
+        query_file_path: The file path to the aggregate query definition.
+        data_file_path: The file path to save the downloaded data.
+    """
+    with open(query_file_path) as f:
+        headcount_json = f.read()
+    query_dto = AggregationQueryExecutionDTO.from_json(headcount_json)
+
+    api_client = ApiClient(config, 'Accept', 'text/csv')
+    query_api = DataQueryApi(api_client)
+
+    logger.info(f"Executing query: {query_file_path}")  # Combined messages
+    response = query_api.aggregate_without_preload_content(query_dto)
+    if not 200 <= response.status <= 299:
+        raise Exception(f"API request failed with status code {response.status}: {response.data.decode()}")
+
+    os.makedirs(os.path.dirname(data_file_path), exist_ok=True)
+    with open(data_file_path, mode='w') as f:
+        f.write(response.data.decode())
+    logger.info(f"Data saved to temp file: {data_file_path}")  # Combined messages
+
+
+def upload_data(config: Configuration, data_file_path: str):
+    """
+    Uploads data using the Data API.
+    Args:
+        config: The API configuration object.
+        data_file_path: The file path to the data to be uploaded.
+    """
+    api_client = ApiClient(config, 'Content-Type', 'application/octet-stream')
+    upload_api = DataUploadApi(api_client)
+
+    with open(data_file_path, 'rb') as f:
+        data = f.read()
+
+    filename = os.path.basename(data_file_path)
+    logger.info(f"Uploading data: {filename}")  # Kept for visibility
+    upload_api.v1_data_upload_files_filename_put(filename=filename, body=data)
+
+
+def main():
+    try:
+        logger.info("Time Machine started.")
```

Comment:
Not sure I like calling this sample app Time Machine. It doesn't allow traveling in time. I'd like to reserve that name for when we're providing a sample that allows you to change the query API to run off a historical DV. That would be more like a time machine. This app is more like snapshot recording. Check with PdM if they have an alternative name they're ok with

Reply:
Wayne agreed to call it Snapshot Recorder, and if there's a better name we'll rename it.